Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,190 @@
+            self.badTransition(event: event)
+
+        case (.registeredNeverActivated, .register):
+            self.badTransition(event: event)
+
+        case (.registeredNeverActivated, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { pipeline in
+                pipeline.fireChannelUnregistered0()
+            }
+
+        // origin: .registered
+        case (.registered, .activate):
+            return self.doStateTransfer(newState: .activated, promise: promise) { pipeline in
+                pipeline.fireChannelActive0()
+            }
+
+        case (.registered, .deactivate):
+            self.badTransition(event: event)
+
+        case (.registered, .register):
+            self.badTransition(event: event)
+
+        case (.registered, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { pipeline in
+                pipeline.fireChannelUnregistered0()
+            }
+
+        // origin: .activated
+        case (.activated, .activate):
+            self.badTransition(event: event)
+
+        case (.activated, .deactivate):
+            return self.doStateTransfer(newState: .registered, promise: promise) { pipeline in
+                pipeline.fireChannelInactive0()
+            }
+
+        case (.activated, .register):
+            self.badTransition(event: event)
+
+        case (.activated, .close):
+            return self.doStateTransfer(newState: .closed, promise: promise) { pipeline in
+                pipeline.fireChannelInactive0()
+                pipeline.fireChannelUnregistered0()
+            }
+
+        // origin: .error
+        case (.closed, _):
+            self.badTransition(event: event)
+        }
+    }
+
+    // MARK: private API
+    @inline(__always)
+    private func badTransition(event: Event) -> Never {
+        fatalError("illegal transition: state=\(self.currentState), event=\(event)")
+    }
+
+    @inline(__always)
+    private mutating func doStateTransfer(newState: State, promise: EventLoopPromise<Void>?, _ callouts: @escaping (ChannelPipeline) -> Void) -> (() -> Void) {
+        self.currentState = newState
+
+        let pipeline = self.channelPipeline
+        return {
+            promise?.succeed(result: ())
+            callouts(pipeline)
+        }
+    }
+
+    // MARK: convenience properties
+    internal var isActive: Bool {
+        assert(self.eventLoop.inEventLoop)
+        switch self.currentState {
+        case .neverRegistered, .registeredNeverActivated, .registered, .closed:
+            return false
+        case .activated:
+            return true
+        }
+    }
+
+    internal var isRegistered: Bool {
+        assert(self.eventLoop.inEventLoop)
+        switch self.currentState {
+        case .neverRegistered, .closed:
+            return false
+        case .registeredNeverActivated, .registered, .activated:
+            return true
+        }
+    }
+
+    internal var isOpen: Bool {
```

Comment:
This is such a weird property. I wonder if we really need it.

Reply:
@Lukasa we use it all over the place to check if a channel has not been closed. Sure, I could introduce a `isClosed` property but then we'd need to `guard !...isClosed` which gives us the inversion always...